extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchBufferAggregator.java

```diff
@@ -87,12 +81,27 @@ public Object get(ByteBuffer buf, int position)
   //Note that this is not threadsafe and I don't think it needs to be
   private Union getUnion(ByteBuffer buf, int position)
   {
-    Union union = unions.get(position);
-    if (union == null) {
-      Memory mem = new MemoryRegion(nm, position, maxIntermediateSize);
-      union = (Union) SetOperation.wrap(mem);
-      unions.put(position, union);
+    Int2ObjectMap<Union> unionMap = unions.get(buf);
+    Union union = unionMap != null ? unionMap.get(position) : null;
+    if (union != null) {
+      return union;
+    }
+    return createNewUnion(buf, position, true);
+  }
+
+  private Union createNewUnion(ByteBuffer buf, int position, boolean isWrapped)
+  {
+    NativeMemory nm = getNativeMemory(buf);
+    Memory mem = new MemoryRegion(nm, position, maxIntermediateSize);
+    Union union = isWrapped
+                  ? (Union) SetOperation.wrap(mem)
+                  : (Union) SetOperation.builder().initMemory(mem).build(size, Family.UNION);
+    Int2ObjectMap<Union> unionMap = unions.get(buf);
```

Comment:
@akashdw 
> results were straightforward that is why I didn't explain it.

"Analyzing benchmarks" is not comparing two numbers. Read this: https://shipilev.net/talks/jvmls-July2014-benchmarking.pdf.

> #4034 suggests ComputeIfAbsent should be used for java 8 map and that is why I performed this benchmark on arbitrary map (Int2ObjectOpenHashMap). If not the case please do not force it in other places.

I didn't force and didn't plan to force that on collections from fastutil because we are losing the specialization benefit. That's what I complained about recently: https://github.com/vigna/fastutil/issues/57. Clarified in #4034.

> I also changed your benchmark to use HashMap instead of IdentityHashMap 

So how this benchmark is relevant, if we use `IdentityHashMap` in real code? To prove that `computeIfAbsent()` is bad? Since your results are counter-intuitive I suspect some problem with the benchmark itself. Deeper analysis could prove or disprove that for this particular pattern of usage of `HashMap.computeIfAbsent()` (which is BTW not identical to the pattern in the real code, apart from that there is a different Map class), but I'm not going to do this, because it is irrelevant.

> getUnion() is a hot spot

Then you can make it much more faster (compared to hand-written/computeIfAbsent difference) if you get rid of those IdentitiyHashMaps altogether. Your objections here: https://github.com/druid-io/druid/pull/4071#discussion_r106768296 are for possible-future-BufferAggregator-users who violate the contract of calling aggregate() with the same ByteBuffer (serially). You can either
 - Add a sanity check `buffer == expectedBuffer` and throw Exception otherwise
 - Or keep IdentityHashMap, but add fields cachedBB and cachedUnionMap and add something like `if (bb == cachedBB) return cachedUnionMap.get(position)` in the beginning of `getUnion()`.

Also, a bigger threat for me seems that someone occasionally passes in reality the same buffer, but different BB object, e. g. `bb.duplicate()`. And it won't result in exception, it will just copy unions in place. I. e. it will be a completely silent performance bug. Neither IdentityHashMap nor HashMap will solve this, you should track down `bb.array()` or `DirectBuffer.address()`